### PR TITLE
Add reactive attributes

### PR DESCRIPTION
**Pull request by davidbrochart**
This PR implements reactive widget attributes. Attributes are stored in a `YMap` in the `Widget`. This means that attributes can be created (and synced) dynamically, there is no need to declare them ahead of time. A local widget (typically in the kernel) can set attributes that a remote widget (typically in the browser) is not even aware of, and the remote widget will discover them at run-time. But of course nothing prevents declaring attributes ahead of time, as class attributes.

cc @martinRenou @trungleduc @SylvainCorlay

### CODE REVIEW

**Comment by davidbrochart** [comment]
Yes, you could still create whatever structure you want from `_ydoc`, and it should still work. But you wouldn't be able to use the feature described above. I think this is fine, as it would allow to do simple things easily, and allow for more complex documents.

File: ypywidgets/ypywidgets.py [hunk 1]
```diff
@@ -17,22 +17,38 @@ class Widget:
...
-        self.name = name
-        self.ydoc = Y.YDoc()
-        if open_comm:
-            self.comm_id = uuid4().hex
-            self.comm = comm.create_comm(
-                comm_id=self.comm_id,
-                target_name=self.name,
+        self._name = name
+        self._ydoc = Y.YDoc()
+        self._attrs = self._ydoc.get_map("attrs")
```
Comment (trungleduc) [inline ypywidgets/ypywidgets.py:26]: Should we name it `_attrs` to prevent naming collision?

**Comment by davidbrochart** [comment]
I created [reacttrs](https://github.com/davidbrochart/reacttrs), which consists of reactive attributes extracted out from [Textual](https://github.com/Textualize/textual/blob/373fc95fc1a5650fd864053e4153eb2d6b38f087/src/textual/reactive.py). I added an [on_set](https://github.com/davidbrochart/reacttrs/blob/7a363623071ff26554554ef0d5e743d632f7102e/reacttrs/reactive.py#L136-L137) callback which is used here to send Y updates over Comm when setting an attribute. The API is the following (see Textual's [documentation](https://textual.textualize.io/guide/reactivity)):
```py
class MyWidget(Widget):
    foo = reactive("my_value")
    bar = reactive(my_function)  # for dynamic defaults

    def watch_foo(self, old, new):
        # called when foo changes
        ...

    def validate_foo(self, value):
        # called when setting foo
        ...
        return value

    def compute_foo(self):
        # called when getting foo
        ...
        return value
```
The (optional) `watch_`, `validate_` and `compute_` methods are followed by the attribute name. It could be interesting to follow this API.
Thoughts?

**Comment by martinRenou** [comment]
Cool!

> The (optional) watch_, validate_ and compute_ methods are followed by the attribute name. It could be interesting to follow this API.

I really like that you separate watch/validate/compute. Could validate and compute be implemented using the official setter and getter for attributes?

I like the syntactic sugar of using decorators in Traitlets for such things as validators and watchers. Maybe watch could be a decorator as well?

**Comment by davidbrochart** [comment]
> I really like that you separate watch/validate/compute.

I didn't do that, Textual does: https://textual.textualize.io/guide/reactivity.

> Could validate and compute be implemented using the official setter and getter for attributes?

Mmm not sure to see what you mean, they are used in [get](https://github.com/davidbrochart/reacttrs/blob/5a44e5f3eaff3dc335dceeb57f12ef9a4ad951cb/reacttrs/reactive.py#L106-L112) and [set](https://github.com/davidbrochart/reacttrs/blob/5a44e5f3eaff3dc335dceeb57f12ef9a4ad951cb/reacttrs/reactive.py#L120-L124).

> I like the syntactic sugar of using decorators in Traitlets for such things as validators and watchers. Maybe watch could be a decorator as well?

Well it's just another syntactic sugar here, with a bit more magic since you don't need the decorator. I personally prefer Textual's API over Traitlets'.

**Comment by martinRenou** [comment]
> Mmm not sure to see what you mean, they are used in 

I mean using official:

```python
@property
def foo(self):
    return ...

@property.setter
def foo(self, value):
    ...
```

> with a bit more magic since you don't need the decorator

That's what I don't like about it, explicit is better than implicit. That's too magical.

**Comment by davidbrochart** [comment]
I don't think `@property` and `@foo.setter` work with the descriptor protocol.

> explicit is better than implicit

But Textual is better than Traitlets :smile:

**Comment by davidbrochart** [comment]
This PR works with the example in [yjs-widgets](https://github.com/davidbrochart/yjs-widgets/tree/66e83149cb34c7456072e03da5b26286695565ac#usage), and this in a notebook:
```py
from ypywidgets import Widget, reactive

class MyWidget1(Widget):
    foo1 = reactive("foo1")
    bar1 = reactive("bar1")

w1 = MyWidget1()
#
w1
# widget output shows "hello from my widget" (set from the frontend)
w1.foo1 = "foo2"
# shows "foo2" in the widget output
w1.foo1 = "foo3"
# shows "foo3" in the widget output
w1.bar1
# prints "'abc'" (set from the frontend)
```

**Comment by davidbrochart** [comment]
If we were to change the magic `watch_foo` method, what would be the preferred API?
I don't really like Traitlets' API, which is also kind of magic with the attribute name as a string:
```py
class Foo(HasTraits):
    foo = Int()

    @observe('foo')
    def _observe_foo(self, change):
        ...
```

**Comment by davidbrochart** [comment]
Let's merge for now, we can always revisit the API in the future.

**Pull request merged**